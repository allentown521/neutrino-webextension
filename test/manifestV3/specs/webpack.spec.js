const _ = require('lodash')
const Neutrino = require('neutrino/Neutrino')
const react = require('@neutrinojs/react')
const webext = require('../../../lib')

describe('webpack', () => {
  const NODE_ENV = process.env.NODE_ENV
  afterAll(() => {
    process.env.NODE_ENV = NODE_ENV
  })

  describe('development', () => {
    beforeEach(() => {
      process.env.NODE_ENV = 'development'
    })

    describe('snapshots', () => {
      test('entries without polyfill', () => {
        testEntriesSnapshot()
      })

      test('entries with polyfill', () => {
        testEntriesSnapshot(true)
      })
    })

    test('should not have webext plugin', () => {
      const neutrino = new Neutrino(optionsFixture())
      neutrino.use(react())
      neutrino.use(webext())
      expect(neutrino.config.plugins.has('webext')).toBeFalsy()
    })

    test('should add global and entry setup files', () => {
      const neutrino = new Neutrino({
        root: '/project',
        source: '/project/src',
        mains: {
          index: {
            entry: 'index',
            webext: {
              type: 'action',
              setup: '/entry-setup-path'
            }
          }
        }
      })
      neutrino.use(react())
      neutrino.use(
        webext({
          setup: '/global-setup-path'
        })
      )
      expect(neutrino.config.entry('index').values()).toMatchSnapshot()
    })

    test('should resolve relative setup path', () => {
      const neutrino = new Neutrino({
        root: '/project',
        source: '/project/src',
        mains: {
          index: {
            entry: 'index',
            webext: {
              type: 'action',
              setup: 'entry-setup-path'
            }
          }
        }
      })
      neutrino.use(react())
      neutrino.use(
        webext({
          setup: 'global-setup-path'
        })
      )
      expect(neutrino.config.entry('index').values()).toMatchSnapshot()
    })

    describe('wextentry cli option', () => {
      test('other entries except background should be removed', () => {
        const bakArgv = process.argv
        process.argv = ['yarn', 'start', '--wextentry', 'popup']
        const neutrino = new Neutrino(optionsFixture())
        neutrino.use(react())
        neutrino.use(webext())
        expect(neutrino.config.entryPoints.has('background')).toBeTruthy()
        expect(neutrino.config.entryPoints.has('popup')).toBeTruthy()
        expect(neutrino.config.plugins.has('html-popup')).toBeTruthy()
        expect(neutrino.config.entryPoints.has('page')).toBeFalsy()
        expect(neutrino.config.plugins.has('html-page')).toBeFalsy()
        expect(neutrino.config.entryPoints.has('pageless')).toBeFalsy()
        expect(neutrino.config.plugins.has('html-pageless')).toBeFalsy()
        expect(neutrino.config.entryPoints.has('content1')).toBeFalsy()
        expect(neutrino.config.plugins.has('html-content1')).toBeFalsy()
        expect(neutrino.config.entryPoints.has('content2')).toBeFalsy()
        expect(neutrino.config.plugins.has('html-content2')).toBeFalsy()

        process.argv = bakArgv
      })
    })
  })

  describe('production', () => {
    beforeEach(() => {
      process.env.NODE_ENV = 'production'
    })

    describe('snapshots', () => {
      test('entries without polyfill(should not be injected)', () => {
        testEntriesSnapshot()
      })

      test('entries with polyfill(should also not be injected)', () => {
        testEntriesSnapshot(true)
      })
    })

    test('should have webext plugin', () => {
      const neutrino = new Neutrino(optionsFixture())
      neutrino.use(react())
      neutrino.use(webext())
      expect(neutrino.config.plugins.has('webext')).toBeTruthy()
    })

    test('should not inject polyfill if polyfill options is off', () => {
      const neutrino = new Neutrino(optionsFixture())
      neutrino.use(react())
      neutrino.use(webext())
      expect(neutrino.config.plugins.has('html-background')).toBeFalsy()
      expect(neutrino.config.plugins.has('html-pageless')).toBeFalsy()
      expect(neutrino.config.plugins.has('html-content1')).toBeFalsy()
      expect(neutrino.config.plugins.has('html-content2')).toBeFalsy()
    })

    test('should inject polyfill if polyfill options is on', () => {
      const neutrino = new Neutrino(optionsFixture())
      neutrino.use(react())
      neutrino.use(
        webext({
          polyfill: 'polyfill-path',
          template: 'template-path'
        })
      )
      expect(neutrino.config.plugins.has('html-background')).toBeFalsy()
      expect(neutrino.config.plugins.has('html-pageless')).toBeFalsy()
      expect(neutrino.config.plugins.has('html-content1')).toBeFalsy()
      expect(neutrino.config.plugins.has('html-content2')).toBeFalsy()
    })
  })
})

function testEntriesSnapshot (polyfill) {
  const neutrino = new Neutrino(optionsFixture())
  neutrino.use(react())
  neutrino.use(polyfill ? webext({ polyfill: true }) : webext())
  expect(
    _.mapValues(neutrino.config.entryPoints.entries(), entry => {
      return entry.values()
    })
  ).toMatchSnapshot()
}

function optionsFixture () {
  return {
    root: '/project',
    source: '/project/src',
    mains: {
      background: {
        entry: 'background',
        webext: {
          type: 'background'
        }
      },
      popup: {
        entry: 'popup',
        webext: {
          type: 'browser_action'
        }
      },
      page: {
        entry: 'page',
        webext: {
          type: 'page'
        }
      },
      pageless: {
        entry: 'pageless',
        webext: {
          type: 'pageless'
        }
      },
      content1: {
        entry: 'content1',
        webext: {
          type: 'content_scripts',
          setup: 'content1/__dev__/setup',
          manifest: {
            matches: ['<all_urls>']
          }
        }
      },
      content2: {
        entry: 'content2',
        webext: {
          type: 'content_scripts',
          manifest: {
            matches: ['https://github.com/crimx/neutrino-webextension'],
            run_at: 'document_start',
            match_about_blank: true,
            all_frames: true
          }
        }
      }
    }
  }
}
