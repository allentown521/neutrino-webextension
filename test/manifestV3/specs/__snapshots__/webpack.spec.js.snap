// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`webpack development should add global and entry setup files 1`] = `
Array [
  "webextensions-emulator/dist/core",
  "/global-setup-path",
  "/entry-setup-path",
  "/project/src/index",
]
`;

exports[`webpack development should resolve relative setup path 1`] = `
Array [
  "webextensions-emulator/dist/core",
  "/project/global-setup-path",
  "/project/src/entry-setup-path",
  "/project/src/index",
]
`;

exports[`webpack development snapshots entries with polyfill 1`] = `
Object {
  "background": Array [
    "webextensions-emulator/dist/background",
    "/project/src/background",
  ],
  "content1": Array [
    "webextensions-emulator/dist/core",
    "/project/src/content1/__dev__/setup",
    "/project/src/content1",
  ],
  "content2": Array [
    "webextensions-emulator/dist/core",
    "/project/src/content2",
  ],
  "page": Array [
    "webextensions-emulator/dist/core",
    "/project/src/page",
  ],
  "pageless": Array [
    "webextensions-emulator/dist/core",
    "/project/src/pageless",
  ],
  "popup": Array [
    "webextensions-emulator/dist/core",
    "/project/src/popup",
  ],
}
`;

exports[`webpack development snapshots entries without polyfill 1`] = `
Object {
  "background": Array [
    "webextensions-emulator/dist/background",
    "/project/src/background",
  ],
  "content1": Array [
    "webextensions-emulator/dist/core",
    "/project/src/content1/__dev__/setup",
    "/project/src/content1",
  ],
  "content2": Array [
    "webextensions-emulator/dist/core",
    "/project/src/content2",
  ],
  "page": Array [
    "webextensions-emulator/dist/core",
    "/project/src/page",
  ],
  "pageless": Array [
    "webextensions-emulator/dist/core",
    "/project/src/pageless",
  ],
  "popup": Array [
    "webextensions-emulator/dist/core",
    "/project/src/popup",
  ],
}
`;

exports[`webpack production snapshots entries with polyfill(should also not be injected) 1`] = `
Object {
  "background": Array [
    "/home/crimx/code/neutrino-webextension/node_modules/webpack-target-webextension/lib/background.js",
    "/project/src/background",
  ],
  "content1": Array [
    "/project/src/content1",
  ],
  "content2": Array [
    "/project/src/content2",
  ],
  "page": Array [
    "/project/src/page",
  ],
  "pageless": Array [
    "/project/src/pageless",
  ],
  "popup": Array [
    "/project/src/popup",
  ],
}
`;

exports[`webpack production snapshots entries without polyfill(should not be injected) 1`] = `
Object {
  "background": Array [
    "/home/crimx/code/neutrino-webextension/node_modules/webpack-target-webextension/lib/background.js",
    "/project/src/background",
  ],
  "content1": Array [
    "/project/src/content1",
  ],
  "content2": Array [
    "/project/src/content2",
  ],
  "page": Array [
    "/project/src/page",
  ],
  "pageless": Array [
    "/project/src/pageless",
  ],
  "popup": Array [
    "/project/src/popup",
  ],
}
`;
